Using mermaid for diagrams blog (#255)
kubernetes/contributor-site: content/en/blog/2021/blog-mermaid.md

---
title: "Improve your documentation with diagrams using Mermaid.js"
linkTitle: "Improve your documentation with diagrams using Mermaid.js"
date: 2021-11-01
---
By Chris Metz and Tim Bannister

---

Topics covered in this blog.

> ![blog outline](/img/mermaid-blog-outline.svg)

Reviewer: nit:
```suggestion
![Why are diagrams useful for documentation? Use mermaid.js / examples](/img/mermaid-blog-outline.svg)
```

* **Share diagram URLs**. The live editor generates a URL for each saved diagram. You can share this link with colleagues to collaborate on diagram creation and editing.

* **Generate svg image files**. If your site doesn't support Mermaid, but can handle svg image files as most do, then just ask the live editor to generate an svg file. __HINT: we use svg files in this blog because the contributor-site does not support Mermaid.__

You can choose from different formats including flowchart, sequence and state. Mermaid works out of the box with the [K8s.io](https://kubernetes.io/) website. 

The Mermaid documentation is excellent - simple text, multiple examples and diagrams included!

---

## What do I need to start using Mermaid?

You need your favorite markdown editor, live editor access and familiarity with Hugo. 

## Examples

The first diagram is shown in [What is Ingress?](https://kubernetes.io/docs/concepts/services-networking/ingress/#what-is-ingress)

> ![ingress-svg](/img/ingress-figure.svg)

Reviewer: ```suggestion
![Ingress](/img/ingress-figure.svg)
```

 end
 classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
 classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
 classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
 class ingress,service,pod1,pod2 k8s;
 class client plain;
 class cluster cluster;
```

---
<br>

Did you notice that some of the diagram elements are colored blue? That is accomplished using a `classDef` statement to define a class of style attributes, and a `class` statement to define the elements to apply that class to.

In the code for the diagram above, we have examples of both: 
```
classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff; // defines style for the k8s class
class ingress,service,pod1,pod2 k8s; // k8s class is applied to elements ingress, service, pod1 and pod2.
```

You can even style an individual element in your diagram using the style statement.
```mermaid
style box1 fill:#FF0000,stroke:333,stroke-width:3px // applies style to box1
```

---
<br>

The next diagram appears in [Pod Topology Spread Constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#example-one-topologyspreadconstraint).

> ![pod-spread](/img/pod-topo-constraint-02.svg)

Reviewer: Here's an example of using a figure shortcode
```suggestion
{{< figure src="/img/pod-topo-constraint-02.svg" title="Example: One TopologySpreadConstraint " caption="Example: One TopologySpreadConstraint (rendered by Mermaid)" link="https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#example-one-topologyspreadconstraint" >}}
```

Author: Got it, will replace with shortcode.

code:
```mermaid
graph TB
subgraph "zoneB"
   p3(Pod) --> n3(Node3)
   n4(Node4)
end
subgraph "zoneA"
   p1(Pod) --> n1(Node1)
   p2(Pod) --> n2(Node2)
end
 
classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
class n1,n2,n3,n4,p1,p2,p3 k8s;
class zoneA,zoneB cluster;
```

---
<br>

> A [recent talk at Kubecon](https://kccncna2021.sched.com/event/lV37/beyond-block-diagrams-different-ways-of-understanding-k8s-architecture-kim-schlesinger-digitalocean) suggested the use of sequence diagrams to describe the interactions between different components in the control plane and nodes. Here is an example illustrating the system flow starting with a `kubectl apply` command and ending with container startup on the node.

Reviewer: https://www.youtube.com/watch?v=X40LJM0KuQ8 is a video of that talk.

Author: should I replace kubecon reference with YT reference?

Reviewer 2: I think you should -- the kubecon page has a good description of the talk, but I think if I were sent there I'd just be looking for the YT link

Author: done. tx @nate-double-u

---

## Why are diagrams useful for documentation?

* **Friendly landing spot**. Greeting readers with a page full of text can be intimidating to those new to Kubernetes, software engineering and tech writing.

* **Faster grasp of concepts.** A diagram can serve as a visual roadmap for details covered in the accompanying text.

* **Better retention.** Humans remember pictures better than words.

Most importantly, readers **acquire the confidence to explore further**. "Wow! Okay I get this. What's next? Maybe I can contribute."

There are many drawing tools to choose from. Traditionally, you create a diagram, save it as png or svg file and then embed the file in the documentation. This takes time, skills vary, and complicates collaboration between contributors and reviewers.

> What if we told you there is a package that allows you to generate simple diagrams using a markdown-like syntax?  

Reviewer: ```suggestion
What if we told you there is a package that allows you to generate simple diagrams using a Markdown-like syntax?
```
(and same elsewhere)

---

## Mermaid.js

[Mermaid](https://mermaid-js.github.io/mermaid/#/) is a package for generating diagrams using markdown-ish text in markdown files. Here are some of the features we love about Mermaid:

* **Inline code syntax** you can copy into the page markdown. Indeed, you should treat Mermaid code as documentation text subject to the usual PR processes. 

* **[On-line live editor](https://mermaid-js.github.io/mermaid-live-editor)** to dynamically create and edit diagrams. The live editor displays the code and the rendered diagram. __TIP: You should always use the live editor to draw diagrams.__

* **Share diagram URLs**. The live editor generates a URL for each saved diagram. You can share this link with colleagues to collaborate on diagram creation and editing.

* **Generate svg image files**. If your site doesn't support Mermaid, but can handle svg image files as most do, then just ask the live editor to generate an svg file. __HINT: we use svg files in this blog because the contributor-site does not support Mermaid.__

> You can choose from different formats including flowchart, sequence and state. Mermaid works out of the box with the [K8s.io](https://kubernetes.io/) website. 

Reviewer: ```suggestion
You can choose from different formats including flowchart, sequence and state.
The main [Kubernetes website](https://kubernetes.io/) is set up so that Mermaid works out of the box.
```
